db/hints/manager.cc

```diff
@@ -351,18 +351,15 @@ future<db::commitlog> manager::end_point_hints_manager::add_store() noexcept {
 
             return commitlog::create_commitlog(std::move(cfg)).then([this] (commitlog l) {
                 // add_store() is triggered every time hint files are forcefully flushed to I/O (every hints_flush_period).
-                // When this happens we want to refill _sender's segments only if it has finished with the segments he had before.
-                if (_sender.have_segments()) {
+                // We would like to keep the list of segments up-to-date, so after every flush we append segments
+                // we didn't see before to the queue.
+                return do_with(l.get_segments_to_replay(), [this] (std::vector<sstring>& segs_vec) {
+                    return do_for_each(segs_vec, [this] (sstring& seg) {
+                        _sender.add_segment(std::move(seg));
+                    });
+                }).then([l = std::move(l)] () mutable {
                     return make_ready_future<commitlog>(std::move(l));
-                }
-
-                std::vector<sstring> segs_vec = l.get_segments_to_replay();
-
-                std::for_each(segs_vec.begin(), segs_vec.end(), [this] (sstring& seg) {
-                    _sender.add_segment(std::move(seg));
                 });
-
-                return make_ready_future<commitlog>(std::move(l));
```

Comment:
Again a comment that is on the existing code, not the patch: why is hints manager bothering itself with segments? It should view commitlog as a complete entity, and not bother with its sub-components. Why not create and replay it as a whole?

Reply:
I remember Calle said that the commitlog object was designed to be write-only. During its lifetime, it only writes commitlog segments and deletes them at appropriate times when memtables are flushed.

Commitlog segments are supposed to be read only in an exceptional case after the node crashed, so the API is limited - there is the `commitlog::get_segments_to_replay` function for getting a list of segments from a previous execution, and a single `commitlog::read_log_file` for reading from log files one at a time.

Those design decisions were made with the regular commitlog in mind, so without modifying the commitlog API, hints manager has to use the commitlog in the way it does now.

Reply:
We can modify the commitlog API if it makes sense.